plugins/registration.ts

```diff
@@ -17,53 +18,115 @@ interface RegistrationFeatures {
     /** The username used to register the client to the server. */
     username?: string;
 
+    /** The password for the user account associated with the username field. */
+    password?: string;
+
     /** The realname used to register the client to the server. */
     realname?: string;
 
-    /** The password used to connect the client to the server. */
-    password?: string;
+    /** Optional server password. */
+    serverPassword?: string;
+
+    /** The authentication method to use. Defaults to NickServ if omitted.
+     * * `NickServ` - Non-standard nickserv authentication.
+     * * `sasl` - SASL PLAIN auth. Errors out if SASL fails.
+     * * `saslThenNickServ` - Try SASL PLAIN, but fallback to NickServ if it fails.
+     */
+    authMethod?: "NickServ" | "sasl" | "saslThenNickServ";
   };
   state: {
     user: User;
   };
 }
 
+function* chunks(str: string, n: number): Generator<string, void> {
+  for (let i = 0; i < str.length; i += n) {
+    yield str.slice(i, i + n);
+  }
+}
+
 export default createPlugin(
   "registration",
   [cap, nick, register],
 )<RegistrationFeatures>((client, options) => {
-  const { nick, username = nick, realname = nick, password } = options;
+  const {
+    nick,
+    username = nick,
+    realname = nick,
+    serverPassword,
+    password,
+  } = options;
+
+  const authMethod = options.authMethod || "NickServ";
   client.state.user = { nick, username, realname };
 
   const sendRegistration = () => {
-    if (password !== undefined) {
-      client.pass(password);
-    }
+    if (serverPassword) client.pass(serverPassword);
     client.nick(nick);
     client.user(username, realname);
   };
 
-  // Sends capabilities and registers once connected.
+  const tryNickServ = () => {
+    client.send("PRIVMSG", "NickServ", `identify ${username} ${password}`);
```

Comment:
What you did is right, but just to share an info: in a case like that, you can also use `client.privmsg("NickServ", ...)` if you want. The only thing to do, in order to make `.privmsg()` available, is to inject the required plugin:

```ts
import privmsg from "./privmsg.ts";

export default createPlugin(
  "registration",
  [cap, nick, register, privmsg], // <-- "privmsg" added
)<RegistrationFeatures>((client, options) => {
  // ...
  client.privmsg(...) // <-- is available now
```

Reply:
Aha, I wondered why I wasn't able to call .privmsg(). This is a really neat architecture honestly.